internal/agent/device/hook/manager.go

```diff
@@ -2,270 +2,169 @@ package hook
-	m.onBeforeRemove = beforeRemoveMap
-	m.onAfterCreate = afterCreateMap
-	m.onAfterUpdate = afterUpdateMap
-	m.onAfterRemove = afterRemoveMap
-	m.onBeforeReboot = beforeRebootMap
-	m.onAfterReboot = afterRebootMap
-	m.initialized.Store(true)
-	return nil
+	return m.executeActions(ctx, actions, actionCtx)
 }
 
-func (m *manager) setError(path string, err error) {
-	m.mu.Lock()
-	defer m.mu.Unlock()
+func (m *manager) loadAndMergeActions(hookType api.DeviceLifecycleHookType) ([]api.HookAction, error) {
+	actionsMap := map[string][]api.HookAction{}
+	// Read actions from the read-only hooks directory (/usr/lib/flightctl/hooks.d/${hookType}/*.yaml)
+	err := m.loadActions(actionsMap, filepath.Join(ReadOnlyConfigDir, HooksDropInDirName, strings.ToLower(string(hookType)), "*.yaml"))
 	if err != nil {
-		m.errors[path] = err
-	} else {
-		delete(m.errors, path)
+		return []api.HookAction{}, err
 	}
-}
-
-func (m *manager) getActionsForPath(path string, actions ActionMap) []ActionHook {
-	m.mu.Lock()
-	defer m.mu.Unlock()
-	return append(actions[path], actions[filepath.Dir(path)]...)
-}
-
-func (m *manager) submitBackgroundJob(path string, actions ActionMap) {
-	actionList := m.getActionsForPath(path, actions)
-	if len(actionList) == 0 {
-		return
+	// Overlay actions from the user-writable hooks directory (/etc/flightctl/hooks.d/${hookType}/*.yaml)
+	err = m.loadActions(actionsMap, filepath.Join(UserWritableConfigDir, HooksDropInDirName, strings.ToLower(string(hookType)), "*.yaml"))
+	if err != nil {
+		return []api.HookAction{}, err
 	}
-	m.backgroundJobs <- func(ctx context.Context) {
-		m.runActionList(ctx, path, actionList)
+	// Sort files containing actions in lexical order, then flatten actionMap into a list of actions in that order
+	keyList := make([]string, 0, len(actionsMap))
+	for k := range actionsMap {
+		keyList = append(keyList, k)
 	}
-}
-
-func (m *manager) Run(ctx context.Context) {
-	for {
-		select {
-		case job, ok := <-m.backgroundJobs:
-			if !ok {
-				m.log.Warn("Hooks manager: jobs channel closed")
-				return
-			}
-			job(ctx)
-		case <-ctx.Done():
-			m.log.Debug("Hooks manager: context closed")
-			return
-		}
+	sort.Strings(keyList)
+	actions := []api.HookAction{}
+	for _, k := range keyList {
+		actions = append(actions, actionsMap[k]...)
 	}
+	return actions, nil
 }
 
-func (m *manager) runActionList(ctx context.Context, path string, actionHooks []ActionHook) {
-	if len(actionHooks) == 0 {
-		return
-	}
-	var errs []error
-	for _, actionHook := range actionHooks {
-		if err := actionHook.OnChange(ctx, path); err != nil {
-			m.log.Errorf("error while running hook for path %s: %+v", path, err)
-			errs = append(errs, fmt.Errorf("failed to run hook for path %s: %w", path, err))
+func (m *manager) loadActions(actionsMap map[string][]api.HookAction, actionFilesGlob string) error {
+	actionFiles, _ := filepath.Glob(m.reader.PathFor(actionFilesGlob))
+	for _, f := range actionFiles {
+		contents, err := os.ReadFile(f)
```

Comment:
do we need to consider fileio.Reader?

Reply:
Tried that originally. Problem is `filepath.Glob()` above already needs the rootDir prefixed, so the list of matching files does, too. Using fileio.Reader would the double-prefix the rootDir.